src/bsoncxx/builder/list.hpp

```diff
@@ -29,148 +30,119 @@ namespace builder {
     ///
-    /// @see bsoncxx::types::bson_value::view.
+    /// @param init
+    ///     the initializer list used to construct the BSON document
     ///
-    operator bson_value::view() {
-        return view();
+    document(initializer_list_t init) : _is_document{true} {
+        for (auto&& kvp : init) {
+            _core.key_view(kvp.first);
+            kvp.second._is_document ? _core.append(kvp.second._core.view_document())
+                                    : _core.append(kvp.second._value);
+        }
+    }
+
+    operator bsoncxx::document::value() {
+        return this->extract();
+    }
+
+    bsoncxx::document::value extract() {
+        return _core.extract_document();
     }
 
     ///
-    /// Provides a view of the underlying BSON value.
+    /// Appends a bsoncxx::builder::document.
     ///
-    /// @see bsoncxx::types::bson_value::view.
+    /// @exception
+    ///    Throws a bsoncxx::exception if the document is malformed.
     ///
-    bson_value::view view() {
-        return val.view();
+    document& operator+=(document rhs) {
+        if (!rhs._is_document)
+            throw bsoncxx::exception{error_code::k_internal_error};
+        _core.concatenate(rhs.extract());
+        return *this;
     }
 
    private:
-    bson_value::value val;
-
-    friend class document;
-    friend class array;
-
-    list(initializer_list_t init, bool type_deduction, bool is_array) : val{nullptr} {
-        std::stringstream err_msg{"cannot construct document"};
-        bool valid_document = false;
-        if (type_deduction || !is_array) {
-            valid_document = [&] {
-                if (init.size() % 2 != 0) {
-                    err_msg << " : must be list of key-value pairs";
-                    return false;
-                }
-                for (size_t i = 0; i < init.size(); i += 2) {
-                    auto t = (begin(init) + i)->val.view().type();
-                    if (t != type::k_utf8) {
-                        err_msg << " : all keys must be string type. ";
-                        err_msg << "Found type=" << to_string(t);
-                        return false;
-                    }
-                }
-                return true;
-            }();
-        }
-
-        if (valid_document) {
-            core _core{false};
-            for (size_t i = 0; i < init.size(); i += 2) {
-                _core.key_owned(std::string((begin(init) + i)->val.view().get_string().value));
-                _core.append((begin(init) + i + 1)->val);
-            }
-            val = bson_value::value(_core.extract_document());
-        } else if (type_deduction || is_array) {
-            core _core{true};
-            for (auto&& ele : init)
-                _core.append(ele.val);
-            val = bson_value::value(_core.extract_array());
-        } else {
-            throw bsoncxx::exception{error_code::k_unmatched_key_in_builder, err_msg.str()};
-        }
-    }
+    bool _is_document{false};
```

Comment:
Do we still need this flag now that this class always builds documents?

Reply:
Yes. Though I guess the name is a bit misleading. This boolean checks whether this class simply wraps a `bson_value::value` in anticipation of being a part of a larger document. I renamed it to `_is_value` and updated the logic.

Reply:
Edit: Removed this.